src/boot/bootctl-status.c

```diff
@@ -503,6 +504,141 @@ int verb_status(int argc, char *argv[], void *userdata) {
         return r;
 }
 
+static int ref_file(Hashmap *known_files, const char *fn, const char *root)
+{
+        const char *p, *k = NULL;
+        int r, n;
+
+        assert(known_files);
+        if (!fn)
+                return -EINVAL;
+
+        p = prefix_roota(root, fn);
+
+        n = PTR_TO_INT(hashmap_get2(known_files, p, (void**)&k));
+        if (!k)
+                r = hashmap_put(known_files, strdup(p), INT_TO_PTR(1));
+        else
+                r = hashmap_update(known_files, p, INT_TO_PTR(++n));
+
+        if (r < 0)
+                return log_error_errno(r, "Failed to add %s to hashmap: %m", p);
+
+        return n;
+}
+
+static int deref_file(Hashmap *known_files, const char *fn, const char *root)
+{
+        const char *p;
+        int n;
+
+        assert(known_files);
+        if (!fn)
+                return -EINVAL;
+
+        p = prefix_roota(root, fn);
+
+        n = PTR_TO_INT(hashmap_get(known_files, p));
+        if (n > 0) {
+                int r = hashmap_update(known_files, p, INT_TO_PTR(--n));
+                if (r < 0)
+                        return log_error_errno(r, "failed to update %s", p);
+        }
+        /* don't remove key even if count is zero as that would
+         * produce a memory leak: There is no API to free the key.
+         * Ony the destructor can do that. */
+
+        return n;
+}
+
+static int deref_unlink_file(Hashmap *known_files, const char *fn, const char *root)
+{
+        _cleanup_free_ char *path = NULL;
+        int r;
+        r = deref_file(known_files, fn, root);
+        if (r == 0) {
+                r = chase_symlinks_and_access(fn, root, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, F_OK, &path, NULL);
+                if (r < 0)
+                        return r;
+                if (unlink(path) < 0)
+                        return log_error_errno(errno, "failed to remove \"%s\": %m", path);
+                log_info("Removed %s", path);
+        }
+        return r;
+}
+
+static int count_known_files(const BootConfig *config, Hashmap **known_files)
+{
+        assert(config);
+        assert(known_files);
+
+        *known_files = hashmap_new(&string_hash_ops);
```

Comment:
we have this rule in systemd that we should initialize all return params on success, and none on failure. Here you write directly to the return parameter even though the function might still fail later.

Hence, please allocate the hashmap in a local variable first, and only once you know for sure the function cannot fail anymore, move it into the return parameters.

also see coding style docs

Reply:
should use path_hash_ops, not string_hash_ops